### PR TITLE
Catch more unquantifiable elements

### PR DESCRIPTION
**Pull request by natecook1000**
This adds start/end anchors ^ and $, groups that form zero-width assertions, and option-changing groups.

### CODE REVIEW

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by natecook1000** [comment]
@swift-ci Please test

File: Tests/RegexTests/ParseTests.swift [hunk 1]
```diff
@@ -2569,6 +2565,11 @@ extension RegexTests {
     diagnosticTest(#"\Z??"#, .notQuantifiable)
     diagnosticTest(#"\G*?"#, .notQuantifiable)
     diagnosticTest(#"\z+?"#, .notQuantifiable)
+    diagnosticTest(#"^*"#, .notQuantifiable)
+    diagnosticTest(#"$?"#, .notQuantifiable)
+    diagnosticTest(#"(?=a)+"#, .notQuantifiable)
+    diagnosticTest(#"(?i)*"#, .notQuantifiable)
+    diagnosticTest(#"(?i:)?"#, .notQuantifiable)
```
Comment (hamishknight) [inline Tests/RegexTests/ParseTests.swift:2572]: PCRE seems to allow quantified change matching option groups (just not the isolated case `(?i)`), do we need to ban them? Would you have to nest the group e.g `(?:(?i:...))?` instead?

**Review by hamishknight — COMMENTED** [review]
Thanks! I'm not sure whether or not we want to ban quantified change matching option groups, but otherwise this LGTM!

**Comment by natecook1000** [comment]
@swift-ci Please test

**Review by hamishknight — APPROVED** [review]

**Pull request merged**